modules/data.atmosphere/R/download.NOAA_GEFS_downscale.R

```diff
@@ -0,0 +1,341 @@
+  
+  time = seq(from = start_date + lubridate::hours(6), to = end_date, by = "6 hour") 
+  forecasts = matrix(ncol = length(noaa_data)+ 2, nrow = 0)
+  colnames(forecasts) <- c(cf_var_names, "timestamp", "NOAA.member")
+  
+  index = matrix(ncol = length(noaa_data), nrow = 64)
+  for(i in 1:21){
+    rm(index)
+    index = matrix(ncol = length(noaa_data), nrow = 64)
+    for(j in 1:length(noaa_data)){
+      index[,j] <- noaa_data[[j]][i,]
+      colnames(index) <- c(cf_var_names) 
+      index <- as.data.frame(index)
+    }
+    index$timestamp <- as.POSIXct(time)
+    index$NOAA.member <- rep(i, times = 64)
+    forecasts <- rbind(forecasts, index)
+  }
+  
+  forecasts <- forecasts %>% tidyr::drop_na()
+  #forecasts$timestamp <- as.POSIXct(rep(time, 21))
+  forecasts$wind_speed <- sqrt(forecasts$eastward_wind^ 2 + forecasts$northward_wind^ 2)
+  
+  ### Downscale state variables 
+  gefs_hour <- PEcAn.data.atmosphere::downscale_spline_to_hourly(df = forecasts, VarNamesStates = c("air_temperature", "wind_speed", "specific_humidity", "precipitation_flux", "air_pressure"))
+  
+  ## convert longwave to hourly (just copy 6 hourly values over past 6-hour time period)
+  nonSW.flux.hrly <- forecasts %>%
+    dplyr::select(timestamp, NOAA.member, surface_downwelling_longwave_flux_in_air) %>%
+    PEcAn.data.atmosphere::downscale_repeat_6hr_to_hrly() %>% dplyr::group_by_at(c("NOAA.member", "timestamp")) %>% 
+    dplyr::summarize(surface_downwelling_longwave_flux_in_air = mean(surface_downwelling_longwave_flux_in_air))
+  
+  ## downscale shortwave to hourly
+  time0 = min(forecasts$timestamp)
+  time_end = max(forecasts$timestamp)
+  ShortWave.ds = PEcAn.data.atmosphere::downscale_ShortWave_to_hrly(forecasts, time0, time_end, lat = 45.805925, lon = -90.07961, output_tz= "UTC")%>% 
+    dplyr::group_by_at(c("NOAA.member", "timestamp")) %>% 
+    dplyr::summarize(surface_downwelling_shortwave_flux_in_air = mean(surface_downwelling_shortwave_flux_in_air))
+  
+  
+  joined<-  dplyr::inner_join(gefs_hour, nonSW.flux.hrly, by = c("NOAA.member", "timestamp"))
+  
+  joined <- dplyr::inner_join(joined, ShortWave.ds, by = c("NOAA.member", "timestamp")) %>% 
+    dplyr::distinct() %>% 
+    dplyr::mutate(surface_downwelling_shortwave_flux_in_air = ifelse(surface_downwelling_shortwave_flux_in_air < 0, 0, surface_downwelling_shortwave_flux_in_air),
+                  specific_humidity = ifelse(specific_humidity <0, 0, specific_humidity),
+                  air_temperature = ifelse(air_temperature > 320, NA, air_temperature),
+                  air_temperature = ifelse(air_temperature < 240, NA, air_temperature),
+                  precipitation_flux = ifelse(precipitation_flux < 0, 0, precipitation_flux),
+                  surface_downwelling_longwave_flux_in_air = ifelse(surface_downwelling_longwave_flux_in_air < 0, NA, surface_downwelling_longwave_flux_in_air),
+                  wind_speed = ifelse(wind_speed <0, 0, wind_speed)) %>%
```

Comment:
Especially since we're using `dplyr` anyway, I would use `dplyr::if_else` throughout here because it is stricter about enforcing type-safety. You will have to chance `NA` to the more explicit `NA_real_` when you do that (because the default type for `NA` is`logical`).